### PR TITLE
refactor: remove for loops from weighted dense matrix creation (#1483)

### PR DESCRIPTION
**Pull request by schochastics**
This PR removes (nested) for loops for dense weighted matrices. Specifically:

- `get.adjacency.dense` when `attr != NULL`
- `graph.incidence.dense` when `weighted = TRUE`
- `get.incidence.dense` when `attr != NULL`

### CODE REVIEW

**Comment by maelle** [comment]
Thank you!! I'll start by adding some tests (from examples) for `as_adjacency_matrix()` and the other functions impacted by this PR in another/other PR(s), then we can rebase this. I have to work on tests anyway.

**Comment by maelle** [comment]
I found a related issue https://github.com/igraph/rigraph/issues/1137 cc @szhorvat

File: R/conversion.R [hunk 1]
```diff
@@ -183,61 +177,22 @@ get.adjacency.dense <- function(graph, type = c("both", "upper", "lower"),
...
-        for (i in seq(length.out = ecount(graph))) {
-          e <- ends(graph, i, names = FALSE)
-          res[e[1], e[2]] <- exattr[i]
-          if (e[1] != e[2]) {
-            res[e[2], e[1]] <- exattr[i]
-          }
-        }
-      }
-    }
+    res <- as.matrix(get.adjacency.sparse(graph, type = type, attr = attr, names = names))
```
Comment (krlmlr) [inline R/conversion.R:190]: Should we add a comment here that this is actually faster than hand-rolling a specialized implementation?

File: R/incidence.R [hunk 2]
```diff
@@ -101,46 +100,27 @@ graph.incidence.dense <- function(incidence, directed, mode, multiple,
...
-          }
-        }
-      }
+
+    idx <- which(incidence != 0, arr.ind = TRUE)
+    el <- cbind(idx, incidence[idx])
+
+    el[, 2] <- el[, 2] + n1
+
+    if (!directed || mode == 1) {
```
Comment (krlmlr) [inline R/incidence.R:109]: This is hard to follow because at this stage in the code the meaning of `mode` is lost. Can/should we postpone the conversion of `mode` to numbers to a later stage so that we have more readable code here?

File: R/incidence.R [hunk 3]
```diff
@@ -101,46 +100,27 @@ graph.incidence.dense <- function(incidence, directed, mode, multiple,
...
-            weight[ptr] <- incidence[i, j]
-            ptr <- ptr + 1
-          }
-        }
-      }
+
+    idx <- which(incidence != 0, arr.ind = TRUE)
+    el <- cbind(idx, incidence[idx])
+
+    el[, 2] <- el[, 2] + n1
```
Comment (krlmlr) [inline R/incidence.R:107]: Is this the same as

```r
el <- cbind(idx, incidence[idx] + n1)
```

?

In either case, this needs a bit of explanation in a comment.

File: R/conversion.R [hunk 4]
```diff
@@ -825,15 +825,14 @@ get.incidence.dense <- function(graph, types, names, attr) {
...
-      if (!types[eo[1]]) {
-        res[e[1], e[2]] <- edge_attr(graph, attr, i)
-      } else {
-        res[e[2], e[1]] <- edge_attr(graph, attr, i)
-      }
-    }
+    el <- as_edgelist(graph, names = FALSE)
+    idx <- types[el[, 1]]
+    el[, 1] <- recode[el[, 1]]
+    el[, 2] <- recode[el[, 2]]
```
Comment (krlmlr) [inline R/conversion.R:831]: Can we do

```r
el[] <- recode(...)
```

or similar?

File: R/conversion.R [hunk 5]
```diff
@@ -825,15 +825,14 @@ get.incidence.dense <- function(graph, types, names, attr) {
...
+
+    tmp <- el[idx, 2:1]
+    el[idx, ] <- tmp
```
Comment (krlmlr) [inline R/conversion.R:834]: Do we still need `tmp` ?

**Review by krlmlr — COMMENTED** [review]
Thanks, great! The second commit was hard for me to understand, maybe we could create three PRs instead and review independently?

@maelle has #1539 that might improve coverage here. I'm rebasing, so let's check what the test coverage is, and adapt as needed.

File: R/incidence.R (hunk 2 above)
Comment (schochastics) [reply R/incidence.R:97]: I had the same thought but just kept it as it was before. As far as I can see, the only need for the numeric value is for `R_igraph_biadjacency`. I think having the character value here would definitely make sense.

File: R/incidence.R (hunk 3 above)
Comment (schochastics) [reply R/incidence.R:95]: Yes that is the same. I can refactor a bit and add comments

File: R/conversion.R (hunk 4 above)
Comment (schochastics) [reply R/conversion.R:851]: yes I can adapt that

File: R/incidence.R (hunk 3 above)
Comment (schochastics) [reply R/incidence.R:95]: sorry I was wrong here. It is not the same, which makes it obvious that comments are needed...

**Comment by maelle** [comment]
The tests are erroring, is it the tests' or the code's fault? :sweat_smile:

**Comment by schochastics** [comment]
I will have a look!

**Comment by schochastics** [comment]
cc @maelle 

https://github.com/igraph/rigraph/blob/0d96334876472a8e311efc50297cd5de83dee975/tests/testthat/test-conversion.R#L175-L183

Here is an error in the test. The test graph has multiple edges and the edge weights are aggregated. E.g. the entry [3,3] is supposed to be 5.6+6.0=11.6 and not 6.0 and [4,2] is 6.1+ 3.3+4.3=13.7 and not 4.3. The aggregation makes sense and is invoked by `use.last.ij <- FALSE` in `get.adjacency.sparse()`

The later snapshot errors seem to be related with this error (I am not familiar with `expect_snapshot`). Do you want me to fix the tests in this PR, or do you do this in another branch?

**Comment by maelle** [comment]
Thank you! It'd make sense to fix them in another PR to main to check they're passing for main, then we'll merge and rebase this PR. At least I think it makes sense :sweat_smile:

**Comment by schochastics** [comment]
hmm they currently pass on main, so in the current igraph version, get.adjacency.dense does not aggregate multiple edge weights (gut get.adjacency.sparse does). So I guess if you change this test on main, it will fail. But IMHO this is a bug. dense and sparse should give the same result and both aggregate

**Comment by schochastics** [comment]
closing this and splitting it into three separate PRs
